rasterio/plot.py

```diff
@@ -33,38 +33,69 @@
 logger = logging.getLogger(__name__)
 
 
-def show(source, cmap='gray', with_bounds=True):
+def show(source, cmap='gray', with_bounds=True, ax=None, title=None, **kwargs):
     """Display a raster or raster band using matplotlib.
 
     Parameters
     ----------
-    source : array-like or (raster dataset, bidx)
+    source : array-like, or (raster dataset, bidx) tuple, or raster dataset, 
         If array-like, should be of format compatible with
         matplotlib.pyplot.imshow. If the tuple (raster dataset, bidx),
-        selects band `bidx` from raster.
+        selects band `bidx` from raster.  If raster dataset display the rgb image 
+        as defined in the colorinterp metadata, or default to first band. 
     cmap : str (opt)
         Specifies the colormap to use in plotting. See
         matplotlib.Colors.Colormap. Default is 'gray'.
     with_bounds : bool (opt)
         Whether to change the image extent to the spatial bounds of the image,
         rather than pixel coordinates. Only works when source is
-        (raster dataset, bidx).
+        (raster dataset, bidx) or raster dataset.
+    ax : matplotlib axes (opt)
+        The raster will be added to this axes if passed.
+    title : str, optional
+        Title for the figure.
+    **kwargs : optional keyword arguments
+        These will be passed to the matplotlib imshow method. See full list at:
+        http://matplotlib.org/api/axes_api.html?highlight=imshow#matplotlib.axes.Axes.imshow
     """
+    if plt is None:  # pragma: no cover
+        raise ImportError("Could not import matplotlib")
+
     if isinstance(source, tuple):
         arr = source[0].read(source[1])
         if with_bounds:
-            extent = plotting_extent(source[0])
+            kwargs['extent'] = plotting_extent(source[0])
+    elif  isinstance(source, rasterio._io.RasterReader):
+        if source.count == 1:
+            arr = source.read(1, masked=True)
         else:
-            extent = None
+            try:
+                source_colorinterp = {source.colorinterp(n): n for n in source.indexes}
+                colorinterp = rasterio.enums.ColorInterp
+                rgb_indexes = [source_colorinterp[ci] for ci in 
+                              (colorinterp.red, colorinterp.green, colorinterp.blue)]
+                arr = source.read(rgb_indexes, masked=True)
+                arr = reshape_as_image(arr)
+
+                if with_bounds:
+                    kwargs['extent'] = plotting_extent(source)
+            except KeyError:
+                arr = source.read(1, masked=True)
     else:
         arr = source
-        extent = None
-    if plt is not None:
-        plt.imshow(arr, cmap=cmap, extent=extent)
-        fig = plt.gcf()
-        fig.show()
-    else:  # pragma: no cover
-        raise ImportError("matplotlib could not be imported")
+
+    if ax:
+        ax.imshow(arr, cmap=cmap, **kwargs)
+        if title:
+            ax.set_title(title, fontweight='bold')
```

Comment:
@ColinTalbert If I follow correctly, shouldn't the user set this title before they pass in `ax`?  (having a hard time testing, my virtualenv is not playing nice with matplotlib).

I think the behavior of title should be consistent, rather than contingent on whether or not `ax` is present.